### PR TITLE
chore: implementation of a first functional draft

### PR DESCRIPTION
**Pull request by fraxken**
This PR implement:

- read function (with ability to create a default file if it doesn't exist).
- write function (with complete or partial update).

The only key implemented right now is `version` (i would need at least a second one to be able to test partial update).

### CODE REVIEW

**Review by antoine-coulon — CHANGES_REQUESTED** [review]
Looks good to me in general 👍 There are only few issues about experimental stuff about ESM that should be fixed to avoid runtime errors (such as importing .ts file without supplying .js extension). At the moment, it works in test files because we are supplying `experimental-specifier-resolution=node` in the mocharc.json file but for a external use we probably don't want to require providing this type of experimental flag.

File: src/functions/read.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+   *
+   * @default false
+   */
+  createIfDoesNotExist?: boolean;
+}
+
+export async function read(
+  location = process.cwd(),
+  options: readOptions = Object.create(null)
+): Promise<TR.Result<RC, NodeJS.ErrnoException>> {
```
Comment (tony-go) [inline src/functions/read.ts:29]: Feel free to destructure `Result` too.

File: test/read.spec.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import os from "node:os";
+
+// Import Third-party Dependencies
+import { expect } from "chai";
+
+// Import Internal Dependencies
+import { read } from "../src/index";
+import { generateDefaultRC } from "../src/rc";
+
+describe("read .nodesecurerc", () => {
```
Comment (tony-go) [inline test/read.spec.ts:13]: Maybe could we have the case where the `.nodesecurerc` and returns the config?

File: test/write.spec.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    const writeResult = await write(location, { payload });
+    expect(writeResult.ok).equal(true);
+    expect(writeResult.val).equal(void 0);
+
+    const readResult = await read(location, { createIfDoesNotExist: false });
+    expect(readResult.ok).equal(true);
+    expect(readResult.val).deep.equal(payload);
+  });
+
+  // TODO: partial update test
```
Comment (tony-go) [inline test/write.spec.ts:64]: 👀

**Review by tony-go — CHANGES_REQUESTED** [review]
NEAT! 👍

File: src/functions/read.ts (hunk 1 above)
Comment (fraxken) [reply src/functions/read.ts:29]: Not possible, this is a type and ts-results is CJS

File: test/read.spec.ts (hunk 2 above)
Comment (fraxken) [reply test/read.spec.ts:13]: I did not understand 😂

File: test/write.spec.ts (hunk 3 above)
Comment (fraxken) [reply test/write.spec.ts:64]: Yes i need at least two keys to be able to test partialUpdate (as mentionned in my PR).

File: src/functions/read.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
+// Import Node.js Dependencies
+import path from "node:path";
+import { once } from "node:events";
+
+// Import Third-party Dependencies
+import Config from "@slimio/config";
+import TR from "ts-results";
+
+// Import Internal Dependencies
+import schema from "../schema/nodesecurerc.json";
```
[the remark this answers is not in the document]
Comment (fraxken) [reply src/functions/read.ts:10]: Yes good catch, going to be a problem 👍

**Comment by fraxken** [comment]
Room for improvement possible:
- Add TSD test for types
- Enhance readOptions & writeOptions interfaces (found a way to build a proper type with type-fest or a custom type).

**Comment by fraxken** [comment]
Maybe we need to add `@nodesecure` dependencies as optionalDependencies ?

```ts
"dependencies": {
  "@nodesecure/i18n": "^1.2.1",
  "@nodesecure/js-x-ray": "^4.2.0",
  "@nodesecure/vuln": "^1.5.0",
  "@slimio/config": "^1.0.1",
  "ts-results": "^3.3.0",
  "type-fest": "^2.11.0"
}
```

Payload still need discussion but overall the PR is ready to land.

**Pull request merged**